File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """

        # Brute force approach: 
        # - Create a new 2D array and then do the rest of the opeations
        
        # Optimal approach: first column of original array = reverse of transpose first row of rotated array

        # transpose an array
        m = len(matrix) # no. of rows

        for i in range(m):
            for j in range(i):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

        # reverse the rows
        for i in range(m):
            matrix[i].reverse()

File: 0048-rotate-image/0048-rotate-image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    
    m = matrix.length

    for (let i = 0; i < m; i++){
        for (let j = 0; j < i; j++){
            let element = matrix[i][j]
            matrix[i][j] = matrix[j][i]
            matrix[j][i] = element
        }
    }

    for (let i = 0; i < m; i++){
        matrix[i].reverse()
    }
};